fdbserver/storageserver.actor.cpp

```diff
@@ -8724,12 +8759,120 @@ bool fetchKeyCanRetry(const Error& e) {
 	case error_code_commit_proxy_memory_limit_exceeded:
 	case error_code_storage_replica_comparison_error:
 	case error_code_unreachable_storage_replica:
+	case error_code_bulkload_task_failed: // for fetchKey based bulkload
 		return true;
 	default:
 		return false;
 	}
 }
 
+ACTOR Future<BulkLoadFileSet> bulkLoadFetchKeyValueFileToLoad(StorageServer* data,
+                                                              std::string dir,
+                                                              BulkLoadTaskState bulkLoadTaskState) {
+	ASSERT(bulkLoadTaskState.getLoadType() == BulkLoadType::SST);
+	TraceEvent(SevInfo, "SSBulkLoadTaskFetchSSTFile", data->thisServerID)
+	    .setMaxEventLength(-1)
+	    .setMaxFieldLength(-1)
+	    .detail("BulkLoadTask", bulkLoadTaskState.toString())
+	    .detail("Dir", abspath(dir));
+	state double fetchStartTime = now();
+	// Download data file from fromRemoteFileSet to toLocalFileSet
+	state BulkLoadFileSet fromRemoteFileSet = bulkLoadTaskState.getFileSet();
+	state BulkLoadFileSet toLocalFileSet = wait(bulkLoadDownloadTaskFileSet(
+	    bulkLoadTaskState.getTransportMethod(), fromRemoteFileSet, dir, data->thisServerID));
+	// Do not need byte sampling locally in fetchKeys
+	const double duration = now() - fetchStartTime;
+	const int64_t totalBytes = bulkLoadTaskState.getTotalBytes();
+	TraceEvent(SevInfo, "SSBulkLoadTaskFetchSSTFileFetched", data->thisServerID)
+	    .setMaxEventLength(-1)
+	    .setMaxFieldLength(-1)
+	    .detail("BulkLoadTask", bulkLoadTaskState.toString())
+	    .detail("Dir", abspath(dir))
+	    .detail("LocalFileSet", toLocalFileSet.toString())
+	    .detail("Duration", duration)
+	    .detail("TotalBytes", totalBytes)
+	    .detail("Rate", duration == 0 ? -1.0 : (double)totalBytes / duration);
+	return toLocalFileSet;
+}
+
+ACTOR Future<Void> tryGetRangeForBulkLoad(PromiseStream<RangeResult> results,
+                                          KeyRange keys,
+                                          size_t totalKeyCount,
+                                          std::string dataPath) {
+	try {
+		// TODO(BulkLoad): what if the data file is empty but the totalKeyCount is not zero
+		state std::unique_ptr<IRocksDBSstFileReader> reader = newRocksDBSstFileReader();
+		reader->open(abspath(dataPath));
+		state size_t keyIndex = 0;
+		state Key lastKey;
+		state KeyValue kv;
+		state RangeResult rep;
+		loop {
+			rep.clear();
+			while (reader->hasNext()) {
```

Comment:
Could we set a bound in the sst reader and skip the unwanted data? 
Ref https://github.com/facebook/rocksdb/blob/880f85a162c2363c6ee102cf5fcaa126bcbc2010/include/rocksdb/sst_file_reader.h#L31

Reply:
Let's discuss this offline.

Reply:
As discussed offline, I changed the implementation using SST file bounded read.